core/embed/trezorhal/rdi.c

```diff
@@ -70,46 +91,54 @@ __attribute__((target("thumb"))) void rdi_handler(uint32_t uw_tick) {
 
     asm volatile(
         "ldr r0, %0;"  // r0 = delay
-        "add r0, $3;"  // r0 += 3
+        "add r0, $6;"  // r0 += 6
         "loop:"
         "subs r0, $3;"  // r0 -= 3
         "bhs loop;"     // if (r0 >= 3): goto loop
-        // loop ((delay // 3) + 1) times
-        // one extra loop ensures that branch predictor learns the loop
+        // loop ((delay // 3) + 2) times
+        // two extra loops ensures that branch predictor learns the loop
         // every loop takes 3 ticks
         // r0 == (delay % 3) - 3
-        "lsl r0, $1;"      // r0 *= 2
-        "add r0, $4;"      // r0 += 4
-        "rsb r0, r0, $0;"  // r0 = -r0
-        // r0 = 2 if (delay % 3 == 0) else 0 if (delay % 3 == 1) else -2 if
-        // (delay % 3 == 2)
-        "add pc, r0;"  // jump (r0 + 2)/2 instructions ahead
-        // jump here if (delay % 3 == 2)
-        "nop;"  // wait one tick
-        // jump here if (delay % 3 == 1)
-        "nop;"  // wait one tick
-        // jump here if (delay % 3 == 0)
+        "add r0, $3;"  // r0 += 3
+        // r0 == delay % 3
+        "and r0, r0, $3;"  // r0 %= 3, make sure that 0 <= r0 < 4
```

Comment:
Isn't it just to make the number posive (higest bit clear)? Not sure if needed.

Reply:
The `and` is a failsafe against fault injection. But what I meant is that the comment doesn't make sense, it should be either:
- `r0 %= 4`
- `r0 &= 3`
Nevertheless, you are right that the `add` isn't actually needed if the table were reshuffled, because `and` would clear the high bits.

Reply:
Typo. Fixed.